packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1602,11 +1602,52 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
   ///  * [Form], which provides an `onWillPop` callback that uses this mechanism.
   ///  * [addScopedWillPopCallback], which adds callback to the list
   ///    checked by [willPop].
+  @Deprecated(
+    'Use registerPopScope or PopScope instead. '
+    'This feature was deprecated after v3.12.0-1.0.pre.',
+  )
   void removeScopedWillPopCallback(WillPopCallback callback) {
     assert(_scopeKey.currentState != null, 'Tried to remove a willPop callback from a route that is not currently in the tree.');
     _willPopCallbacks.remove(callback);
   }
 
+  /// Registers the existence of a [PopScope] widget in the route.
+  ///
+  /// [PopScope] widgets registered in this way will have their
+  /// [PopScope.onPopped] callbacks called when a route is popped or a pop is
+  /// attempted. They will also be able to block pop operations with
+  /// [PopScope.popEnabled] through this route's [popEnabled] method.
+  ///
+  /// See also:
+  ///
+  /// [unregisterPopScope], which performs the opposite operation.
+  void registerPopScope(PopScope widget) {
+    _popScopes.add(widget);
+    _updateSystemNavigator();
+  }
+
+  /// Unregisters a [PopScope] widget in the route's widget subtree.
+  ///
+  /// See also:
+  ///
+  /// [registerPopScope], which performs the opposite operation.
+  void unregisterPopScope(PopScope widget) {
+    _popScopes.remove(widget);
+    _updateSystemNavigator();
+  }
+
+  // Tells the SystemNavigator whether or not a system pop should have an
+  // effect.
+  void _updateSystemNavigator() {
+    final NavigationNotification notification = NavigationNotification(
+      // canPop indicates that the originator of the Notification can handle a
+      // pop. In the case of PopScope, it handles pops when popEnabled is
+      // false. Hence the seemingly backward logic here.
+      canHandlePop: popEnabled() == RoutePopDisposition.doNotPop,
```

Comment:
should this only send when this route is the top-most route? also should the route also be a navigationNotificationListener to in case there is a navigator in the subtree? This way, you can remove all the noise from the navigators in the route beneath.

Reply:
Good point, I will short circuit this when `isCurrent` is not true.

I put a NotificationListener in the Navigator for this purpose so that it can also check the Navigator's `canPop`.

Reply:
If you don't put a navigationNotificationListener here, how would the NotificationListener in the Navigator be able to tell whether a notification is coming from nested a navigator that are burry underneath the top-most route

Reply:
See the logic for this in navigator.dart's NotificationListener ([link to the code](https://github.com/flutter/flutter/pull/120385/files#diff-c2be6c8a787644c1625f577babbdf321bf6416feeac7a838155c9e2633413808R5425), though I'm not sure if the link will work).

Basically it bubbles up the widget tree.  Each Navigator will either allow it to keep bubbling or re-dispatch its own notification if canHandlePop has changed.